backend/app/store/search/internal/buffered_engine.go

```diff
@@ -0,0 +1,296 @@
+
+	err := s.index.Batch(batch)
+	if err != nil {
+		log.Printf("[ERROR] error while indexing batch, %v", err)
+	}
+	for _, notifier := range notifiers {
+		notifier.notifier <- err
+	}
+}
+
+// indexDocumentWorker processes documents from current batch in a loop
+// exit when queueNotifier is closed and decrements shutdownWait counter
+// Note: shutdownWait should be incremented before this call
+func (s *bufferedEngine) indexDocumentWorker() {
+	log.Printf("[INFO] start bleve indexer worker")
+	defer s.shutdownWait.Done()
+
+	tmr := time.NewTimer(s.flushEvery)
+	cont := true
+	for cont {
+		var force bool
+		select {
+		case <-tmr.C:
+			s.indexBatch()
+			tmr.Reset(s.flushEvery)
+		case force, cont = <-s.queueNotifier:
+			s.queueLock.RLock()
+			full := s.docQueue.Len() >= s.flushCount
+			s.queueLock.RUnlock()
+			if force || full {
+				s.indexBatch()
+			}
+		}
+	}
+	log.Printf("[INFO] shutdown bleve indexer worker")
+
+	s.dumpAheadLog()
+}
+
+func (s *bufferedEngine) getAheadLogPath() string {
+	return path.Join(s.indexPath, aheadLogFname)
+}
+
+// dumpDoc writes document to file separeted with \0
+func dumpDoc(f *os.File, doc *service.DocumentComment) error {
+	data, err := json.Marshal(doc)
+	if err != nil {
+		return err
+	}
+	data = append(data, 0x0)
+	_, err = f.Write(data)
+	return err
+}
+
+func (s *bufferedEngine) dumpAheadLog() {
+	var err error
+
+	aheadLogPath := s.getAheadLogPath()
+	if _, errOpen := os.Stat(aheadLogPath); !os.IsNotExist(errOpen) {
+		log.Printf("[ERROR] file %q already exists and would be rewritten", aheadLogPath)
+	}
+
+	f, err := os.Create(filepath.Clean(aheadLogPath))
+	if err != nil {
+		log.Printf("[ERROR] error %v opening log file %q", err, aheadLogPath)
+		return
+	}
+	defer func() {
+		errClose := f.Close()
+		if errClose != nil {
+			log.Printf("[ERROR] error %v closing log file %q", errClose, aheadLogPath)
+		}
+	}()
+
+	s.queueLock.Lock()
+	defer s.queueLock.Unlock()
+
+	notifiers := []*idxFlusher{}
+	// write all unprocessed documents into a file
+	for s.docQueue.Len() > 0 {
+		switch val := s.docQueue.PopFront().(type) {
+		case *service.DocumentComment:
+			if err != nil {
+				// we don't stop processing on error
+				// because we want to collect all waiters to send them an error
+				continue
+			}
+			err = dumpDoc(f, val)
+			if err != nil {
+				log.Printf("[ERROR] error %v writing log file", err)
+			}
+		case *idxFlusher:
+			// we will send error to all waiters, because indexer will not process this documents now
+			notifiers = append(notifiers, val)
+		default:
+			panic(fmt.Sprintf("unknown type %T", val))
+		}
+	}
+
+	for _, notifier := range notifiers {
+		notifier.notifier <- errors.Errorf("indexer closing")
+	}
+}
+
+// Init engine. It loads dumped comments from ahead log saved from buffer on shutdown
+// Return true if engine initialized before, false means cold start
+func (s *bufferedEngine) Init(ctx context.Context) (bool, error) {
+	// TODO(@vdimir) add tests for this part
```

Comment:
I feel that this part need to be tested, but not sure how. I'll try to do something

Reply:
Generally, it's a good idea to create real instance of something and test internal behavior using only public methods. That way it's easy to refactor if needed and easy to replicate in case you'll add new search engine.

Reply:
This part is difficult to test as black box. And for this I've written special test